### PR TITLE
Request for comments: eve FTP logging

### PR DESCRIPTION
**Pull request by jlucovsky**
*Request for comments only*

- Please ignore the code marked guarded with `#if 0`
- Code guarded with `#if 1` will remains
- When an actual PR is submitted, all `#if 0` and `#if 1` guards will be removed
- This PR omits documentation; documentation updates will be included in the actual PR

This changeset is for issue [2315](https://redmine.openinfosecfoundation.org/issues/2315) and implements eve logging for FTP.

I'd appreciate feedback on the log records as well as any other comments. Currently, transaction details are contained within a node termed `current` -- suggestions for (1) an alternate name or (2) how to represent differently are solicited.

Only a limited number of commands are logged -- when a command is logged, the original request, response, and response code are logged in addition to flow information.

There are 2 new event types:
1. ftp
2. ftp-data

## Sample output (event_type ftp)

```
{
  "timestamp": "2019-03-10T15:20:21.104835-0700",
  "flow_id": 1235831968963670,
  "pcap_cnt": 125,
  "event_type": "ftp",
  "src_ip": "2a01:0e34:ee97:b130:8c3e:45ea:5ac6:e301",
  "src_port": 38970,
  "dest_ip": "2001:0700:0100:0118:0000:0000:0000:0047",
  "dest_port": 21,
  "proto": "TCP",
  "tx_id": 0,
  "current": {
    "command": "USER",
    "request": "anonymous",
    "response": "Please specify the password",
    "response_code": "331"
  }
}

{
  "timestamp": "2013-06-17T14:59:53.197616-0700",
  "flow_id": 1328603147152946,
  "pcap_cnt": 1379,
  "event_type": "ftp",
  "src_ip": "192.168.0.13",
  "src_port": 59898,
  "dest_ip": "192.168.0.10",
  "dest_port": 21,
  "proto": "TCP",
  "tx_id": 3,
  "current": {
    "command": "PORT",
    "request": "192,168,0,13,234,22",
    "response": "PORT command successful. Consider using PASV",
    "response_code": "200",
    "dynamic_port": 59926,
    "mode": "active"
  }
}
```

## Sample output (event_type ftp-data)

```
{
  "timestamp": "2019-03-10T15:20:01.500157-0700",
  "flow_id": 888918871503645,
  "parent_id": 1235831968963670,
  "pcap_cnt": 68,
  "event_type": "ftp-data",
  "src_ip": "2a01:0e34:ee97:b130:8c3e:45ea:5ac6:e301",
  "src_port": 53697,
  "dest_ip": "2001:0700:0100:0118:0000:0000:0000:0047",
  "dest_port": 51093,
  "proto": "TCP",
  "tx_id": 0,
  "current": {
    "filename": "uio-app.js",
    "command": "RETR"
  }
}
{
  "timestamp": "2019-03-10T15:20:18.000469-0700",
  "flow_id": 340670592172679,
  "parent_id": 1235831968963670,
  "pcap_cnt": 116,
  "event_type": "ftp-data",
  "src_ip": "2001:0700:0100:0118:0000:0000:0000:0047",
  "src_port": 20,
  "dest_ip": "2a01:0e34:ee97:b130:8c3e:45ea:5ac6:e301",
  "dest_port": 41813,
  "proto": "TCP",
  "tx_id": 0,
  "current": {
    "filename": "index.html",
    "command": "RETR"
  }
}
{
  "timestamp": "2005-12-13T06:01:40.693042-0800",
  "flow_id": 1423204100025132,
  "parent_id": 1328603147152946,
  "event_type": "ftp-data",
  "src_ip": "192.168.0.10",
  "src_port": 20,
  "dest_ip": "192.168.0.13",
  "dest_port": 59926,
  "proto": "TCP",
  "tx_id": 0,
  "current": {
    "filename": "temp.txt",
    "command": "RETR"
  }
}
```

### CODE REVIEW

**Comment by victorjulien** [comment]
Our normal scheme is that an event_type:x goes with a object x. So here that would be 'event_type:"ftp"; "ftp": { ... }

File: src/util-error.h [hunk 1]
```diff
@@ -123,6 +123,7 @@ typedef enum {
     SC_ERR_DEBUG_LOG_GENERIC,
     SC_ERR_UNIFIED_LOG_GENERIC,
     SC_ERR_HTTP_LOG_GENERIC,
+    SC_ERR_FTP_LOG_GENERIC,
```
Comment (victorjulien) [inline src/util-error.h]: only append to this enum to keep id's stable

File: src/app-layer-ftp.c [hunk 2]
```diff
@@ -470,9 +596,27 @@ static int FTPParseRequest(Flow *f, void *ftp_state,
 
     int direction = STREAM_TOSERVER;
     while (FTPGetLine(state) >= 0) {
-        FTPParseRequestCommand(state,
-                               state->current_line, state->current_line_len);
+
+        FTPParseRequestCommand(state, state->current_line, state->current_line_len);
+        if (state->command == FTP_COMMAND_UNKNOWN) {
```
Comment (victorjulien) [inline src/app-layer-ftp.c]: Does this mean we only create txs for selected commands? I'm thinking that we can recognize all commands, even if we don't know them. Then we could just track and log the unknown ones as well.

**Comment by victorjulien** [comment]
@jasonish could you have a look at the record format as well?

File: src/util-error.h (hunk 1 above)
Comment (jlucovsky) [reply src/util-error.h]: Will relocate constant.

File: src/app-layer-ftp.c [hunk 3]
```diff
@@ -378,6 +489,21 @@ static void FtpTransferCmdFree(void *data)
     FTPFree(cmd, sizeof(struct FtpTransferCmd));
 }
 
+static uint32_t copy_command_line(uint8_t *dest, uint8_t *src, uint32_t length, uint32_t max_length)
```
[the remark this answers is not in the document]
Comment (jlucovsky) [reply src/app-layer-ftp.c]: Will rename per suggestion.

File: src/app-layer-ftp.c (hunk 2 above)
Comment (jlucovsky) [reply src/app-layer-ftp.c]: I'll modify things so that all currently defined commands in the `enum FtpRequestCommand` are recognized.

**Comment by jlucovsky** [comment]
Continued in #3822

**Pull request closed without merging**